pkg/operator/configobservation/apiserver/observe_apiserver.go

```diff
@@ -0,0 +1,272 @@
+func observeUserClientCABundle(apiServer *configv1.APIServer, recorder events.Recorder, previouslyObservedConfig map[string]interface{}) (map[string]interface{}, map[string]string, []error) {
+	configMapName := apiServer.Spec.ClientCA.Name
+	if len(configMapName) == 0 {
+		return nil, nil, nil // previously observed resource (if any) should be deleted
+	}
+	// The user managed client CA bundle will be combined with other operator managed client CA bundles (by the target
+	// config controller) into a common client CA bundle managed by the operator. As such, since the user managed client
+	// CA bundle is never explicitly referenced in the kube-apiserver config, the returned observed config will always
+	// be empty.
+	return nil, map[string]string{"user-client-ca": configMapName}, nil
+}
+
+// observeDefaultUserServingCertificate observes user managed Secret containing the default cert info for serving
+// secure traffic.
+func observeDefaultUserServingCertificate(apiServer *configv1.APIServer, recorder events.Recorder, previouslyObservedConfig map[string]interface{}) (map[string]interface{}, map[string]string, []error) {
+	var errs []error
+	servingCertSecretName := apiServer.Spec.ServingCerts.DefaultServingCertificate.Name
+	if len(servingCertSecretName) == 0 {
+		return nil, nil, nil // previously observed config and resources (if any) should be deleted
+	}
+	// generate an observed configuration that will configure the kube-apiserver to use the user managed default serving
+	// cert info instead of the operator managed default serving cert info.
+	observedConfig := map[string]interface{}{}
+	certFile := fmt.Sprint(userServingCertPublicCertFile)
+	if err := unstructured.SetNestedField(observedConfig, certFile, "servingInfo", "certFile"); err != nil {
+		return previouslyObservedConfig, nil, append(errs, err)
+	}
+	keyFile := fmt.Sprint(userServingCertPrivateKeyFile)
+	if err := unstructured.SetNestedField(observedConfig, keyFile, "servingInfo", "keyFile"); err != nil {
+		return previouslyObservedConfig, nil, append(errs, err)
+	}
+	return observedConfig, map[string]string{"user-serving-cert": servingCertSecretName}, errs
+}
+
+// observeNamedCertificates observes user managed Secrets containing TLS cert info for serving secure traffic to
+// specific hostnames.
+func observeNamedCertificates(apiServer *configv1.APIServer, recorder events.Recorder, previouslyObservedConfig map[string]interface{}) (map[string]interface{}, map[string]string, []error) {
+	var errs []error
+
+	observedConfig := map[string]interface{}{}
+
+	namedCertificates := apiServer.Spec.ServingCerts.NamedCertificates
+	if len(namedCertificates) > maxUserNamedCerts {
+		// TODO: This should be validation error, not operator error/event.
+		err := fmt.Errorf("spec.servingCerts.namedCertificates cannot have more than %v entries", maxUserNamedCerts)
+		recorder.Warningf("ObserveNamedCertificatesFailed", err.Error())
+		return previouslyObservedConfig, nil, append(errs, err)
```

Comment:
no need for `previouslyObservedConfig` anymore because `newAPIServerObserver` does that logic now. One argument less here.

Reply:
@sttts there is some edge case that depend on returning `nil` as observedConfig in order to force delete... i would keep this in place because it will require some other (more complicated mechanism...) (iow. not work optimizing)